### PR TITLE
Voting power delegation

### PR DESCRIPTION
**Pull request by Tanya-atatakai**
changes:
1) added form for delegation UI
2) removed "you must join Contribute" from help text 
3) switched to the new contract, partially removed usage of old one
4) added undelegate feature
5) added voting power breakdown

balance formatting states:

| 0 balance      | 27.4 | 328.6k | 18.1M |
| ----------- | ----------- | ----------- | ----------- |
| Screenshot 2024-03-26 at 20 42 40   | Screenshot 2024-03-26 at 20 43 41        | Screenshot 2024-03-26 at 20 44 20        | Screenshot 2024-03-26 at 20 44 05        |

Validation/Success/error states:
| validation      | error | success |
| ----------- | ----------- | ----------- |
| Screenshot 2024-03-26 at 20 45 03      | Screenshot 2024-03-26 at 20 45 53       | Screenshot 2024-03-26 at 20 45 34       |

Voting power breakdown:
| Zero voting power      | Delegated to you |  Delegated to you with list of delegators expanded |   You delegated to someone else |
| ----------- | ----------- | ----------- | ----------- |
| Screenshot 2024-03-28 at 11 47 57      | Screenshot 2024-03-28 at 11 47 26       | Screenshot 2024-03-28 at 11 47 33       | Screenshot 2024-03-28 at 13 02 34       |

### CODE REVIEW

File: components/GlobalStyles/index.jsx [hunk 1]
```diff
@@ -45,27 +48,27 @@ const GlobalStyle = createGlobalStyle`
...
   .mr-12 {
     margin-right: 12px !important;
   }
+  .mr-24 {
+    margin-right: 24px;
+  }
   .mr-48 {
     margin-right: 48px !important;
   }
-  .mt-8 {
```
Comment (Tanya-atatakai) [inline components/GlobalStyles/index.jsx:66]: it's duplicated, removed

**Review by kupermind — APPROVED** [review]

File: components/Tweet/index.jsx [hunk 2]
```diff
@@ -31,11 +30,7 @@ const SocialPosterContainer = styled.div`
 
 const ToProposeTweetText = () => (
   <Text type="secondary">
-    To propose a tweet, you must
-    {' '}
-    <Link href="/members">join Contribute</Link>
-    {' '}
-    and hold at least 100k veOLAS.
+    To propose a tweet, you must hold at least 100k veOLAS.
```
Comment (truemiller) [inline components/Tweet/index.jsx:33]:
```suggestion
    To propose a tweet, you must have at least 100k veOLAS voting power.
```

File: components/VotingPower/DelegateMenu.jsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import { delegate } from './requests';
+
+const { Text, Paragraph } = Typography;
+
+const DelegateMenu = (props) => {
+  const [form] = Form.useForm();
+  const { account } = useHelpers();
+  const { delegatorList } = useDelegatorList(account);
+
+  const [delegateddAddress, setDelegatedAddress] = useState(null);
```
Comment (truemiller) [inline components/VotingPower/DelegateMenu.jsx:21]:
```suggestion
  const [delegatedAddress, setDelegatedAddress] = useState(null);
```

File: components/VotingPower/DelegateMenu.jsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  return (
+    <StyledMenu className="p-12">
+      <Text strong>Your voting power</Text>
+      <Paragraph className="my-8">
+        {formatWeiBalanceWithCommas(props.balance)}
+      </Paragraph>
+      <Divider className="my-8" />
+      <Text strong>You’ve delegated to:</Text>
+      <Paragraph className="my-8">
+        {delegateddAddress ? truncateAddress(delegateddAddress) : 'None'}
```
Comment (truemiller) [inline components/VotingPower/DelegateMenu.jsx:57]:
```suggestion
        {delegatedAddress ? truncateAddress(delegatedAddress) : 'None'}
```

**Review by truemiller — APPROVED** [review]

File: components/VotingPower/DelegateMenu.jsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+  return (
+    <StyledMenu className="p-12">
+      <Text strong>Your voting power</Text>
+      <Paragraph className="my-8">
+        {formatWeiBalanceWithCommas(props.balance)}
+      </Paragraph>
+      <Divider className="my-8" />
+      <Text strong>You’ve delegated to:</Text>
+      <Paragraph className="my-8">
```
Comment (truemiller) [inline components/VotingPower/DelegateMenu.jsx:56]: we should either:
- link to etherscan page of address
- add some some tooltip to the address so the user can see the full address

File: components/VotingPower/utils.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @param {string} balanceInWei
+ * @returns formatted balance with appropriate suffix
+ */
+export const formatWeiBalance = (balanceInWei) => {
+  const balanceInEther = ethers.utils.formatEther(balanceInWei);
+
+  const formatNumberWithSuffix = (number) => {
+    if (number >= 1e9) {
+      return `${(number / 1e9).toFixed(1)}B`;
```
Comment (truemiller) [inline components/VotingPower/utils.js:12]: while this is a nice approach, i feel off about the "rounding up" effect that Number.toFixed(1) has.

ideally we would round down, so the user is not confused as to why they cannot delegate in a scenario where they don't have enough voting power

**Comment by oaksprout** [comment]
image

Ideally the truncation would have:
- 5 characters either side (e.g. 0x12345...12345) for better identification
- title atttribute set such that you can hover over to see full address
- linked to etherscan

File: components/VotingPower/utils.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+ * @param {string} address - The Ethereum address to truncate
+ * @returns {string} The truncated address
+ */
+export const truncateAddress = (address, startLimit = 5, endLimit = 3) => (address
+  ? `${address.substring(0, startLimit)}...${address.substring(
+    address.length - endLimit,
+  )}`
+  : '--');
+
+export const DELEGATE_ERRORS_MAP = {
```
Comment (Tanya-atatakai) [inline components/VotingPower/utils.js:52]: @oaksprout can you please review this mapping for displaying errors - do you want to change anything? and please note that "no balance" means no balance in veOlas, not the voting power - this logic is from DelegateContribute contract

**Comment by oaksprout** [comment]
Didn't we say we'd hide on mobile? @Tanya-atatakai 

image

**Review by oaksprout — APPROVED** [review]

**Review by kupermind — APPROVED** [review]

**Review by truemiller — APPROVED** [review]

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**